### DISCUSSION TITLE
Proposal: freeze the JavaCC-7 and make all new development/improvement on JavaCC-8

### DISCUSSION THREAD

**Discussion by zosrothko** [javacc/javacc#238] [General]
Hi

Since there is legitimate question about the status of JavaCC as an open source project and in order to avoid a dispersion of the efforts made by all contributors (present past and future), I am proposing to freeze the JavaCC-7 and focus new development/improvment/bug fixes on JavaCC-8.

Freezing means that this repository [JavaCC-7](git@github.com:javacc/javacc.git) will be archived and readable and forkable by everybody so that no more pull be requested on it.

Any thougth?

**Comment by vlsi** [comment]
javacc-8 is split into several repositories, and it is super-inconventient for the following reasons:
* cross-repository changes are hard. For instance, 
https://github.com/javacc/javacc/pull/153 was split into javacc-8-core and javacc-8-java. If a change needs updating multiple generators, then it would be much worse.
* cross-repository testing is harder as well
* issues will be split between the repositories, and it is impossible to tell if the issue belongs to -core or -java
* 8.0 changes maven coordinates, so it will be non-trivial for the clients to upgrade

The current java-8 repositories have truncated history. I believe no development on top of -8 can be done before 
https://github.com/javacc/javacc-8-java/issues/2 and https://github.com/javacc/javacc-8-core/issues/1 are fixed

**Reply by martinswanson** [reply]
Beat me to it!

**Reply by zosrothko** [reply]
Splitting javacc-8 into seprate & independant modules was a requirement of @new-javacc 

> cross-repository changes are hard. For instance,
> https://github.com/javacc/javacc/pull/153 was split into javacc-8-core and javacc-8-java. If a change needs updating multiple generators, then it would be much worse.
True butcannot be avoided. When a change is made in core on the AST, all integrations tests must passed for all generators. Even if ytou are only intersted in the java generation, you should check that the cpp generation pass. That's why when using the top javacc-8 pom, maven retest all generatros

> cross-repository testing is harder as well
See previous point

> issues will be split between the repositories, and it is impossible to tell if the issue belongs to -core or -java
Wrong. The issue could be in only core or in both core and one generator. So it is up to the contributor to find out where is issue is.

> 8.0 changes maven coordinates, so it will be non-trivial for the clients to upgrade
Not really. The maven coordinate to be change is only the generator, because each generator has a dependency on the core

**Reply by vlsi** [reply]
@zosrothko , are you sure @new-javacc  requires separate **repositories**?
It does not look like there is such a requirement. Here's the recent comment by new-javacc: https://github.com/javacc/javacc/discussions/238#discussioncomment-3109677

> new-javacc: My idea is someone should be able to work on individual code generators without modifying the core

It looks like a desire to have separate modules rather than separate repositories.

Would you please reference the requirement of having separate **git repositories** so I could understand it better?

**Reply by zosrothko** [reply]
This is due ot the Maven release process. See my previous explanation and this [page](https://maven.apache.org/guides/mini/guide-releasing.html)

**Comment by martinswanson** [comment]
Definitely agree with moving to 1 version ... we don't have the time/resource to maintain 1 let alone 2 versions !

Current repo structure:
- javacc
- javacc-8
- javacc-8-core
- javacc-8-js
- etc

Observations:
- the javacc-8* repos do not include any of the history from the javacc repo, so we are dropping information like stars, forks, dependent projects, etc.
- the javacc-8-core appears to have been created from an older version, for example it omits all the new markup documentation
- the javacc-8-core also contains a lot duplicate information, including all the build scripts.

Suggestions
- is there anything preventing us structuring this as a single repo, with the various modules (core, js, etc) as sub-modules? modules could be still be packaged as separate jars. this could significantly reduce duplication and simplify the build system.
- once refactored as a single repo, could it not be pushed back into the main javacc repo as a new version? effectively promote javacc-8 to be version 8 in the javacc repo? thus preserving the full history.

So you end up with 
- javacc
  - core
  - js
  - etc

I'm not a Git expert (nor Maven / Gradle) but when I look at large projects like Apache Flink they have a vast number of modules all contained within a single repo. 

Perhaps this would be a good opportunity to trial the Gradle build - structure into modules, reduce duplication, remove the chaff, and provide hooks for all the new tests we want ;-p

My 2c

**Reply by zosrothko** [reply]
The requirement of independent modules implies that each module go in its own git repository. This is due to the [Maven release process](https://maven.apache.org/guides/mini/guide-releasing.html). Each module has a <scm> element in its pom that specifies the git repository. Then the delivery process is done using those 2 command
```
mvn release:prepare
mvn release:perform
```
During the execution of those commands, Maven change all XX.YY.ZZ-SNAPSHOT  version to XX.YY.NN with NN be equal usualy to ZZ + 1
and it tags the git repo. Thus, there is no other way to comply to the initial requirement thn having each module be in its own git repo.

With this organization, one can improve the java generator independently of the cpp generator both of them depending on the same core version. And when the core is improved, one have just to redeliver a generator if this latest wants to gain the new feature. If the other generators do not want to use the new core feature, then nothing has to be done.

> I'm not a Git expert (nor Maven / Gradle) but when I look at large projects like Apache Flink they have a vast number of modules all contained within a single repo.
Apache Flink is using external dependencies and the source code of those dependencies is not included as part of Flink in the SCM. The source code of the dependencies are in their own SCM repo!

**Reply by vlsi** [reply]
I'm sure it is solvable in Maven, however, I am even more sure this is a **non-issue** for Gradle.

Do you think "maven release process" justifies the work and the overheads to split the repositories? Do you think it explains that the project loses history, forks/stars, and makes it harder for contributors to clone and work on the project?

**Reply by zosrothko** [reply]
> Do you think "maven release process" justifies the work and the overheads to split the repositories?
Yes, there is no other way to comply the requirement of @new-javacc to have independent released modules and especially the core module. @new-javacc wants for exemple to be able to release a fix in the java generator without releasing the c++, c# or js generator and without releasing a new core version.

> Do you think it explains that the project loses history, forks/stars
I think so because that's not possible to split a git repo in 2. So one should start from a new repo, coping the the part of javacc-7 source code involved in the module. Doing so, one loose the initial .git history. And since the commitId are different, I do not think it is possible to recover the old javacc-7 hostory and inject it into each javacc-8 modules. May be it is possible, but it would require a huge work!

**Reply by vlsi** [reply]
> Yes, there is no other way 

Have you explored the possibility of using Gradle for build and release?
I am sure there's no such limitation with Gradle, and you can release the needed submodules right from a single git repository.

At the same time, I think it would be just fine to always release all the modules. At the end of the day, there's not enough time to support "multiple versions of core for multiple versions of generators", so it is reasonable to support just the latest versions, which suggests there's no harm in releasing all the modules always even in the case they contain no changes.

---

> @new-javacc wants for exemple to be able to release a fix in the java generator without releasing the c++, c# or js generator and without releasing a new core version

Would you please reference that requirement so I could better understand the requirement and the intentions behind it?
For instance, would always releasing all the modules cover the requirement? Why "releasing new core" might be bad?

**Comment by new-javacc** [comment]
I'm not a github expert either but I would like the "core"  - the module that takes a .jj/.jjt file and produces data for the tokenizer and parser - should be separate. My idea is someone should be able to work on individual code generators without modifying the core. Of course we can evolve core say for example add util class to generate good method/function names etc.

**Reply by martinswanson** [reply]
You would have a sub-module for core with it's own build file, and be able to execute build tasks for that module independently of the rest. This also makes it much easier for someone developing a module with a dependency on core to manage the build system. 

Pretty sure we could achieve the same with Maven, but might be a nice reason to have a go at adopting Gradle. 

I would defer to @vlsi to advise as he has a lot more experience.

**Reply by zosrothko** [reply]
Each module can be independently build and test with unit tests in its loca directory. But in case of the core module, if the new development changes the AST or any core API used by one or more generator, all integration tests should pass. That is why the top level javacc-8 pom rebuild everything. The only thing to really care about from the dependency opint of view is the javacc-8/codegen/pom.xml

**Comment by martinswanson** [comment]
@new-javacc Would you not also want to separate javacc, jjtree, and jjdoc? I never understood why these are in the same repo, as they seem to be different applications.

**Reply by new-javacc** [reply]
Main issue they are produced by preprocessing of the jj(t) file(s) and the generated parser (data) has deps on it. So anything that does not generate code is in core.

And also I want core to be a level above or a in a separate sibling repo from codegenerators which are purely consuming the data produce by core.

**Reply by zosrothko** [reply]
That's the target and this is achieved by having a git repo per module

**Reply by vlsi** [reply]
>That's the target and this is achieved by having a git repo per module

git repo per module solves it, however, it becomes hard to maintain.
Have you thought of alternative approaches?

**Reply by zosrothko** [reply]
I am working with maven and git for years. So I choose this way beacause it was a tremendous work to bundle all together. If you have a better approach, let us know.

**Reply by vlsi** [reply]
> If you have a better approach, let us know.

Gradle-based build would handle that quite well. There will be no need to split the repositories, and modules could be organized as folders inside a single repository.
